src/e84_geoai_common/llm/models/claude.py

```diff
@@ -280,4 +268,117 @@ def invoke_model_with_request(self, request: ClaudeInvokeLLMRequest) -> ClaudeRe
 
-    # FUTURE implement tool use
+    def response_to_llm_message(
+        self, response: ClaudeResponse, inference_cfg: LLMInferenceConfig
+    ) -> LLMMessage:
+        def _to_llm_content(
+            index: int, c: ClaudeTextContent | ClaudeImageContent | ClaudeToolUseContent
+        ) -> TextContent | Base64ImageContent | LLMToolUseContent:
+            match c:
+                case ClaudeTextContent():
+                    text = c.text
+                    response_prefix = _config_to_response_prefix(inference_cfg)
+                    if index == 0 and response_prefix:
+                        text = response_prefix + text
+                    return TextContent(text=text)
+                case ClaudeImageContent():
+                    return Base64ImageContent(media_type=c.source.media_type, data=c.source.data)
+                case ClaudeToolUseContent():
+                    return LLMToolUseContent(id=c.id, name=c.name, input=c.input)
+
+        return LLMMessage(
+            role="assistant",
+            content=[_to_llm_content(i, c) for i, c in enumerate(response.content)],
+        )
+
+    @staticmethod
+    def llm_message_to_claude_message(msg: LLMMessage) -> "ClaudeMessage":
+        """Converts the generic LLM Message into a ClaudeMessage."""
+
+        def _handle_content(content: LLMMessageContentType) -> ClaudeMessageContentType:
+            match content:
+                case TextContent():
+                    return ClaudeTextContent(text=content.text)
+                case Base64ImageContent():
+                    return ClaudeImageContent(
+                        source=ClaudeImageSource(media_type=content.media_type, data=content.data)
+                    )
+                case LLMToolUseContent():
+                    return ClaudeToolUseContent(
+                        id=content.id, name=content.name, input=content.input
+                    )
+                case LLMToolResultContent():
+                    return BedrockClaudeLLM.llm_tool_result_to_claude_tool_result(content)
+
+        if isinstance(msg.content, str):
+            content = msg.content
+        else:
+            content = [_handle_content(subcontent) for subcontent in msg.content]
+        return ClaudeMessage(role=msg.role, content=content)
+
+    @staticmethod
+    def llm_tool_to_claude_tool(tool: LLMTool) -> ClaudeTool:
+        """Build a ClaudeTool from an LLMTool.
+
+        If LLMTool.output_model is set, the JSON schema of the output model is
+        injected into the tool's description so that the LLM is aware of it.
+        """
+        if tool.input_model is None:
+            input_schema = cast(dict[str, Any], {"type": "object", "properties": {}})
+        else:
+            input_schema = tool.input_model.model_json_schema()
+
+        description = tool.description
+        if tool.output_model is not None:
+            output_schema = pretty_repr(tool.output_model.model_json_schema())
```

Comment:
Why use `pretty_repr` over `json.dumps(schema, indent=2)`?

Reply:
Because I forgot that `json.dumps()` can do indentation. Fixed.